File: Power Query/400-499/411/PQ_411.R
library(tidyverse)
library(readxl)

path <- "400-499/411/PQ_Challenge_411.xlsx"
input <- read_excel(path, range = "A1:K21")
test <- read_excel(path, range = "M1:S21")

base <- c("Category", "Item", "Qty")
other <- setdiff(names(input), base)

prefixes <- input %>%
  summarise(across(all_of(other), ~ any(.x == "BA"))) %>%
  pivot_longer(everything(), names_to = "column", values_to = "has_ba") %>%
  filter(has_ba) %>%
  mutate(prefix = str_extract(column, "^\\d+")) %>%
  pull(prefix) %>%
  unique()

keep <- c(base, other[str_extract(other, "^\\d+") %in% prefixes])
result <- input %>% select(all_of(keep))

all.equal(result, test)
# TRUE

File: Power Query/400-499/411/PQ_411.py
import pandas as pd

path = "400-499/411/PQ_Challenge_411.xlsx"
input = pd.read_excel(path, usecols="A:K", nrows=20)
test = pd.read_excel(path, usecols="M:S", nrows=20)
test.columns = ["Category", "Item", "Qty", "1 INCL.", "1 BREAKOUT $", "3 INCL.", "3 BREAKOUT $"]

base = ["Category", "Item", "Qty"]
other = input.columns.difference(base, sort=False)
prefixes = {
    column.split()[0]
    for column in other
    if input[column].eq("BA").any()
}
keep = base + [column for column in other if column.split()[0] in prefixes]

result = input[keep]
print(result.equals(test))
# True
